### PR TITLE
Implement `tainted_by_errors` in MIR borrowck, use it to skip CTFE

### LINKED ISSUE #93646: Const generic default ICE: The type checker should prevent reading from a never-written local
Const generic defaults are being stabilized, https://github.com/rust-lang/rust/pull/90207/. But I found an issue that occured earlier in the length expressions of arrays, see https://github.com/rust-lang/rust/issues/72115

### Code

```Rust
struct X<const N: usize = {
    let s: &str; s.len()
}>;
```

### Meta

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (4e8fb743c 2022-02-03)
```

### Error output

```
error: internal compiler error: compiler/rustc_const_eval/src/interpret/eval_context.rs:202:17: The type checker should prevent reading from a never-written local

```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1160:9
stack backtrace:
   0:     0x7fd976ea7ebc - std::backtrace_rs::backtrace::libunwind::trace::hddd5a7e459ce4327
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7fd976ea7ebc - std::backtrace_rs::backtrace::trace_unsynchronized::h69db3e09cfd8bf11
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fd976ea7ebc - std::sys_common::backtrace::_print_fmt::h83b461b46804fdab
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7fd976ea7ebc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2f6e5cd8872f3d74
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7fd976f092dc - core::fmt::write::hd64a00f3787c90bd
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/core/src/fmt/mod.rs:1190:17
   5:     0x7fd976e97383 - std::io::Write::write_fmt::h6347402c88d40325
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/io/mod.rs:1653:15
   6:     0x7fd976eac181 - std::sys_common::backtrace::_print::h989064c3e0b39e3c
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7fd976eac181 - std::sys_common::backtrace::print::h4af8c55ce731a045
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7fd976eac181 - std::panicking::default_hook::{{closure}}::h41ef6f6a9fac6b53
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/panicking.rs:295:22
   9:     0x7fd976eabe3f - std::panicking::default_hook::hbaf11e031e0e4a5f
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/panicking.rs:314:9
  10:     0x7fd977689541 - rustc_driver[7b59b981d4c23d31]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7fd976eaca5b - std::panicking::rust_panic_with_hook::h4788f0bd625b1005
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/panicking.rs:702:17
  12:     0x7fd97874d281 - std[daefdbcf0dca51d0]::panicking::begin_panic::<rustc_errors[b77837aed2b0d3e7]::ExplicitBug>::{closure#0}
  13:     0x7fd97874d1f6 - std[daefdbcf0dca51d0]::sys_common::backtrace::__rust_end_short_backtrace::<std[daefdbcf0dca51d0]::panicking::begin_panic<rustc_errors[b77837aed2b0d3e7]::ExplicitBug>::{closure#0}, !>
  14:     0x7fd978751a2f - std[daefdbcf0dca51d0]::panicking::begin_panic::<rustc_errors[b77837aed2b0d3e7]::ExplicitBug>
  15:     0x7fd97875e0cd - std[daefdbcf0dca51d0]::panic::panic_any::<rustc_errors[b77837aed2b0d3e7]::ExplicitBug>
  16:     0x7fd97875d37d - <rustc_errors[b77837aed2b0d3e7]::HandlerInner>::bug
  17:     0x7fd97875ced0 - <rustc_errors[b77837aed2b0d3e7]::Handler>::bug
  18:     0x7fd9785e36d9 - rustc_middle[e664fb30aec2b69e]::ty::context::tls::with_opt::<rustc_middle[e664fb30aec2b69e]::util::bug::opt_span_bug_fmt<rustc_span[26d1c5a8904e1196]::span_encoding::Span>::{closure#0}, ()>
  19:     0x7fd9785e4830 - rustc_middle[e664fb30aec2b69e]::util::bug::opt_span_bug_fmt::<rustc_span[26d1c5a8904e1196]::span_encoding::Span>
  20:     0x7fd9785e47a6 - rustc_middle[e664fb30aec2b69e]::util::bug::bug_fmt
  21:     0x7fd978e7fcf6 - <rustc_const_eval[55e19033940159c7]::interpret::eval_context::InterpCx<rustc_const_eval[55e19033940159c7]::const_eval::machine::CompileTimeInterpreter>>::run
  22:     0x7fd978ebed71 - rustc_const_eval[55e19033940159c7]::const_eval::eval_queries::eval_to_allocation_raw_provider
  23:     0x7fd978f40a9c - rustc_query_system[59fab566a51c2c24]::query::plumbing::try_execute_query::<rustc_query_impl[ae50953882263c5d]::plumbing::QueryCtxt, rustc_query_system[59fab566a51c2c24]::query::caches::DefaultCache<rustc_middle[e664fb30aec2b69e]::ty::ParamEnvAnd<rustc_middle[e664fb30aec2b69e]::mir::interpret::GlobalId>, core[90c82cfbe27f5d33]::result::Result<rustc_middle[e664fb30aec2b69e]::mir::interpret::value::ConstAlloc, rustc_middle[e664fb30aec2b69e]::mir::interpret::error::ErrorHandled>>>
  24:     0x7fd978fb1ebf - <rustc_query_impl[ae50953882263c5d]::Queries as rustc_middle[e664fb30aec2b69e]::ty::query::QueryEngine>::eval_to_allocation_raw
  25:     0x7fd978ebd9b2 - rustc_const_eval[55e19033940159c7]::const_eval::eval_queries::eval_to_const_value_raw_provider
  26:     0x7fd978f4194b - rustc_query_system[59fab566a51c2c24]::query::plumbing::try_execute_query::<rustc_query_impl[ae50953882263c5d]::plumbing::QueryCtxt, rustc_query_system[59fab566a51c2c24]::query::caches::DefaultCache<rustc_middle[e664fb30aec2b69e]::ty::ParamEnvAnd<rustc_middle[e664fb30aec2b69e]::mir::interpret::GlobalId>, core[90c82cfbe27f5d33]::result::Result<rustc_middle[e664fb30aec2b69e]::mir::interpret::value::ConstValue, rustc_middle[e664fb30aec2b69e]::mir::interpret::error::ErrorHandled>>>
  27:     0x7fd978fb207b - <rustc_query_impl[ae50953882263c5d]::Queries as rustc_middle[e664fb30aec2b69e]::ty::query::QueryEngine>::eval_to_const_value_raw
  28:     0x7fd9793b9b4a - <rustc_middle[e664fb30aec2b69e]::ty::context::TyCtxt>::const_eval_global_id
  29:     0x7fd979d3fa49 - <rustc_middle[e664fb30aec2b69e]::ty::context::TyCtxt>::const_eval_resolve
  30:     0x7fd979ce2391 - <rustc_infer[835d7697e80d8145]::infer::InferCtxt>::const_eval_resolve
  31:     0x7fd979cb6659 - rustc_trait_selection[99ac7681cfb15a86]::traits::const_evaluatable::is_const_evaluatable
  32:     0x7fd9792a1679 - <rustc_trait_selection[99ac7681cfb15a86]::traits::fulfill::FulfillProcessor>::progress_changed_obligations
  33:     0x7fd9792aa834 - <rustc_data_structures[9901de6203b925f4]::obligation_forest::ObligationForest<rustc_trait_selection[99ac7681cfb15a86]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[99ac7681cfb15a86]::traits::fulfill::FulfillProcessor, rustc_data_structures[9901de6203b925f4]::obligation_forest::Outcome<rustc_trait_selection[99ac7681cfb15a86]::traits::fulfill::PendingPredicateObligation, rustc_infer[835d7697e80d8145]::traits::FulfillmentErrorCode>>
  34:     0x7fd97929d198 - <rustc_trait_selection[99ac7681cfb15a86]::traits::fulfill::FulfillmentContext as rustc_infer[835d7697e80d8145]::traits::engine::TraitEngine>::select_all_or_error
  35:     0x7fd978a5eda5 - <rustc_typeck[95242508bb98a5fb]::check::fn_ctxt::FnCtxt>::select_all_obligations_or_error
  36:     0x7fd979796ba7 - <rustc_infer[835d7697e80d8145]::infer::InferCtxtBuilder>::enter::<(), <rustc_typeck[95242508bb98a5fb]::check::inherited::InheritedBuilder>::enter<<rustc_typeck[95242508bb98a5fb]::check::wfcheck::CheckWfFcxBuilder>::with_fcx<rustc_typeck[95242508bb98a5fb]::check::wfcheck::check_type_defn<rustc_typeck[95242508bb98a5fb]::check::wfcheck::check_item_well_formed::{closure#2}>::{closure#0}>::{closure#0}, ()>::{closure#0}>
  37:     0x7fd978b2caae - rustc_typeck[95242508bb98a5fb]::check::wfcheck::check_item_well_formed
  38:     0x7fd978f55d9e - rustc_query_system[59fab566a51c2c24]::query::plumbing::try_execute_query::<rustc_query_impl[ae50953882263c5d]::plumbing::QueryCtxt, rustc_query_system[59fab566a51c2c24]::query::caches::DefaultCache<rustc_span[26d1c5a8904e1196]::def_id::LocalDefId, ()>>
  39:     0x7fd978fb7ee5 - <rustc_query_impl[ae50953882263c5d]::Queries as rustc_middle[e664fb30aec2b69e]::ty::query::QueryEngine>::check_item_well_formed
  40:     0x7fd978af6b74 - <rustc_middle[e664fb30aec2b69e]::hir::map::Map>::par_visit_all_item_likes::<rustc_typeck[95242508bb98a5fb]::check::wfcheck::CheckTypeWellFormedVisitor>
  41:     0x7fd97978b21c - <rustc_session[bcbf86d149cb5e94]::session::Session>::track_errors::<rustc_typeck[95242508bb98a5fb]::check_crate::{closure#5}, ()>
  42:     0x7fd9797707cb - rustc_typeck[95242508bb98a5fb]::check_crate
  43:     0x7fd979523287 - rustc_interface[be440f50dcda62f9]::passes::analysis
  44:     0x7fd9799bb729 - rustc_query_system[59fab566a51c2c24]::query::plumbing::try_execute_query::<rustc_query_impl[ae50953882263c5d]::plumbing::QueryCtxt, rustc_query_system[59fab566a51c2c24]::query::caches::DefaultCache<(), core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>>>
  45:     0x7fd9799f98a5 - rustc_query_system[59fab566a51c2c24]::query::plumbing::get_query::<rustc_query_impl[ae50953882263c5d]::queries::analysis, rustc_query_impl[ae50953882263c5d]::plumbing::QueryCtxt>
  46:     0x7fd9794f980d - <rustc_interface[be440f50dcda62f9]::passes::QueryContext>::enter::<rustc_driver[7b59b981d4c23d31]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>>
  47:     0x7fd9794f781a - <rustc_interface[be440f50dcda62f9]::interface::Compiler>::enter::<rustc_driver[7b59b981d4c23d31]::run_compiler::{closure#1}::{closure#2}, core[90c82cfbe27f5d33]::result::Result<core[90c82cfbe27f5d33]::option::Option<rustc_interface[be440f50dcda62f9]::queries::Linker>, rustc_errors[b77837aed2b0d3e7]::ErrorReported>>
  48:     0x7fd9794fa876 - rustc_span[26d1c5a8904e1196]::with_source_map::<core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>, rustc_interface[be440f50dcda62f9]::interface::create_compiler_and_run<core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>, rustc_driver[7b59b981d4c23d31]::run_compiler::{closure#1}>::{closure#1}>
  49:     0x7fd9794f71fe - rustc_interface[be440f50dcda62f9]::interface::create_compiler_and_run::<core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>, rustc_driver[7b59b981d4c23d31]::run_compiler::{closure#1}>
  50:     0x7fd9794dd302 - std[daefdbcf0dca51d0]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[be440f50dcda62f9]::util::setup_callbacks_and_run_in_thread_pool_with_globals<rustc_interface[be440f50dcda62f9]::interface::run_compiler<core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>, rustc_driver[7b59b981d4c23d31]::run_compiler::{closure#1}>::{closure#0}, core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>>::{closure#0}, core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>>
  51:     0x7fd9794daa19 - <<std[daefdbcf0dca51d0]::thread::Builder>::spawn_unchecked_<rustc_interface[be440f50dcda62f9]::util::setup_callbacks_and_run_in_thread_pool_with_globals<rustc_interface[be440f50dcda62f9]::interface::run_compiler<core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>, rustc_driver[7b59b981d4c23d31]::run_compiler::{closure#1}>::{closure#0}, core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>>::{closure#0}, core[90c82cfbe27f5d33]::result::Result<(), rustc_errors[b77837aed2b0d3e7]::ErrorReported>>::{closure#1} as core[90c82cfbe27f5d33]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  52:     0x7fd976eb8803 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hed49a6c6e767ba83
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/alloc/src/boxed.rs:1854:9
  53:     0x7fd976eb8803 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h14238373d01dc9bc
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/alloc/src/boxed.rs:1854:9
  54:     0x7fd976eb8803 - std::sys::unix::thread::Thread::new::thread_start::hc6dbf73cd24f68cc
                               at /rustc/4e8fb743ccbec27344b2dd42de7057f41d4ebfdd/library/std/src/sys/unix/thread.rs:108:17
  55:     0x7fd976dec609 - start_thread
  56:     0x7fd976d06293 - clone
  57:                0x0 - <unknown>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by compiler-errors**
Putting this up for initial review. The issue that I found is when we're evaluating a const, we're doing borrowck, but doing nothing with the fact that borrowck fails. 

This implements a `tainted_by_errors` field for MIR borrowck like we have in infcx, so we can use that information to return an `Err` during const eval if our const fails to borrowck.

This PR needs some cleaning up. I should probably just use `Result` in more places, instead of `.expect`ing in the places I am, but I just wanted it to compile so I could see if it worked!

Fixes #93646

r? @oli-obk
feel free to reassign

### CODE REVIEW

**Comment by compiler-errors** [comment]
I wonder what other places we could use this tainted field that we generate during MIR borrowck...

File: compiler/rustc_borrowck/src/lib.rs [hunk 1]
```diff
@@ -2299,6 +2308,15 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
     fn is_upvar_field_projection(&self, place_ref: PlaceRef<'tcx>) -> Option<Field> {
         path_utils::is_upvar_field_projection(self.infcx.tcx, &self.upvars, place_ref, self.body())
     }
+
+    pub fn buffer_error(&mut self, t: DiagnosticBuilder<'_>) {
+        self.tainted_by_errors = true;
+        t.buffer(&mut self.errors_buffer);
```
Comment (oli-obk) [inline compiler/rustc_borrowck/src/lib.rs:2314]: Seems a bit fragile to keep exposing the buffer list. Maybe nest it in some structure that contains the tainted field and the errors? Or at least leave a FIXME for it ^^

**Review by oli-obk — CHANGES_REQUESTED** [review]
I think https://github.com/rust-lang/rust/pull/93691/commits/cbec6f88e88f6187ade134a3d6972dfc742dcd82 is too invasive and does not fit how most queries handle errors. Maybe we should just add a `tainted_by_errors` field to `mir::Body`, too?

What do you think?

**Comment by compiler-errors** [comment]
@oli-obk, so I thought about putting it into Body, but it looks like borrowck actually treats Body as immutable in favor of returning a different BorrowCheckResult struct, and I didn't want to do something with interior mutability or reallocating a new Body...

**Comment by oli-obk** [comment]
I was thinking of doing it where you now return `Err`, as we definitely have a mutable body available. But at that point the field on the body becomes fairly useless.

We could instead just check borrowck (and probably typeck, too) in `load_mir`. Then we can remove the redundant check in the const eval queries. There are other checks in those queries that may be good to move to `load_mir`, too

**Comment by compiler-errors** [comment]
> We could instead just check borrowck (and probably typeck, too) in `load_mir`. 

Not opposed to this necessarily, but even if we do borrowck again in `load_mir`, we still need to expose some field like the `tainted_by_errors` I added that reflects the fact an error occurred during that check?

**Comment by oli-obk** [comment]
> > We could instead just check borrowck (and probably typeck, too) in `load_mir`.
> 
> Not opposed to this necessarily, but even if we do borrowck again in `load_mir`, don't we still need to expose some field like the `tainted_by_errors` I added that reflects the fact an error occurred during that check?

Oh yes, we need all your borrowck changes , only now the mir queries don't change ^^

**Comment by compiler-errors** [comment]
oh sure, let me see if i can do that then.

File: compiler/rustc_borrowck/src/lib.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_borrowck/src/lib.rs:2314]: Not sure what you mean by this. 

This method on MirBorrowckCtxt is just convenience to both buffer the diagnostic (which we are already doing before this diff) and set tainted_by_errors to true, since I didn't want to have to duplicate setting that boolean (or calling a method to set that boolean) in like the 20 spots I introduced this helper.

**Comment by compiler-errors** [comment]
@oli-obk: Yeah, I didn't need to `Result`-ify all those MIR queries actually. This is much simpler. 

There are two places I do a borrowck call (`eval_to_allocation_raw_provider` and `InterpCx::load_mir`), but those are also two places we're currently also checking for typeck errors. Not sure if it's significant to have both of them, and I can try to deduplicate those if you want.

File: compiler/rustc_borrowck/src/lib.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_borrowck/src/lib.rs:2314]: Sorry. What I mean is that the previous way is still possible. So people may add invocations of `.buffer` on the diagnostic and thus keep forgetting to set the tainted flag. If we hide the buffer as a private field of a struct we can avoid people accessing it directly.

It's more of a concern for careless impls in the future than a problem with the PR.

File: compiler/rustc_middle/src/query/mod.rs [hunk 2]
```diff
@@ -833,13 +833,14 @@ rustc_queries! {
     /// additional requirements that the closure's creator must verify.
     query mir_borrowck(key: LocalDefId) -> &'tcx mir::BorrowCheckResult<'tcx> {
         desc { |tcx| "borrow-checking `{}`", tcx.def_path_str(key.to_def_id()) }
-        cache_on_disk_if(tcx) { tcx.is_typeck_child(key.to_def_id()) }
+        cache_on_disk_if { true }
```
Comment (oli-obk) [inline compiler/rustc_middle/src/query/mod.rs:836]: Why is this necessary? 

I'm going to do a perf run,? but maybe we just need to check the tainted flag here, too

File: compiler/rustc_middle/src/mir/query.rs [hunk 3]
```diff
@@ -214,6 +214,7 @@ pub struct BorrowCheckResult<'tcx> {
     pub concrete_opaque_types: VecMap<OpaqueTypeKey<'tcx>, Ty<'tcx>>,
     pub closure_requirements: Option<ClosureRegionRequirements<'tcx>>,
     pub used_mut_upvars: SmallVec<[Field; 8]>,
+    pub tainted_by_errors: bool,
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/query.rs:217]: Please make it an option like in the typeck results .this will make it easier for the in-flight PRs that make our error handling more solid

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 4]
```diff
@@ -287,6 +287,9 @@ pub fn eval_to_allocation_raw_provider<'tcx>(
             if let Some(error_reported) = tcx.typeck_opt_const_arg(def).tainted_by_errors {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:287]: A bit more than planned for this PR, but if you want to try doing this cleanup in this PR, I'm not going to complain ;)

Basically remove the entire logic in the `is_local` if block and move whatever is missing in `load_mir` to it.

We can also guard the checks in `load_mir` with the is_local check, as any mir we load from dependencies must be error free

Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:287]: That said, feel free not to do this in this PR .we can merge it soon in any case

File: compiler/rustc_middle/src/query/mod.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/query/mod.rs:836]: I was getting ICEs in incremental tests complaining that the MIR body for a fn was stolen, since we're now calling borrowck which depends on unoptimized MIR which is probably gone if we do const eval late in the compilation process...

Not sure if only caching if the body if it is tainted works, since we're unconditionally calling borrowck before const eval.

File: compiler/rustc_borrowck/src/lib.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_borrowck/src/lib.rs:2314]: oh yeah, good point, I can hide buffer in a harder-to-access field to prevent people accidentally not setting tainted with it too.

I can't just set tainted when we drain the buffer later because DiagnosticBuilder::buffer sometimes just emits the error directly instead of buffering it depending on some settings.

File: compiler/rustc_middle/src/query/mod.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/query/mod.rs:836]: Perf does show the additional query result serialization, so we should try to do something here.

Comment (oli-obk) [reply compiler/rustc_middle/src/query/mod.rs:836]: It's probably not worth it, but we could try adding a query that returns just an `Option<ErrorReported>` and does all the processing of typeck, borrowck and const qualifs. Then we can only cache that on disk and it should rarely get recomputed. But: new query so probably overkill, but worth an experiment?

Comment (compiler-errors) [reply compiler/rustc_middle/src/query/mod.rs:836]: Yeah, I'm not exactly sure what we can do to limit `cache_on_disk` just to when it's needed, since it seems like we need to cache it for any body that has a promoted const, see the failing query stack trace:

```
query stack during panic:
#0 [mir_borrowck] borrow-checking `lit_test::lit_test`
#1 [eval_to_allocation_raw] const-evaluating + checking `lit_test::lit_test::promoted[1]`
#2 [eval_to_allocation_raw] const-evaluating + checking `lit_test::lit_test::promoted[1]`
#3 [optimized_mir] optimizing MIR for `lit_test::lit_test`
#4 [collect_and_partition_mono_items] collect_and_partition_mono_items
```

... but we don't know if a body has any promoted consts until after borrowck.

so I might try the query experiment you suggested. I'll ping you for a perf run when I put it up, since I don't have super powers.

Comment (compiler-errors) [reply compiler/rustc_middle/src/query/mod.rs:836]: > Actually, we do know it at borrowck time, so we could encode that info in the borrowck result.

Oh lol, I thought we did borrowck on `mir_built`, whoopsy.

> When we force mir borrowck and then steal the mir, we own the mir, so we can mutate it and change its tainted field.

So you're suggesting we do add the tainted field to `mir::Body` then, instead of `BorrowCheckerResult`?

Comment (oli-obk) [reply compiler/rustc_middle/src/query/mod.rs:836]: You need it on both, as (as you reminded me) you can't mutate the mir from borrowck. But at the query that steals the mir after forcing the borrowck query, you can just call borrowck instead of forcing it and copy the tainted field over to the mir body you just stole

File: src/test/ui/impl-trait/issues/issue-78722.rs [hunk 5]
```diff
@@ -8,6 +8,7 @@ struct Bug {
     V1: [(); {
         fn concrete_use() -> F {
             async {}
+            //~^ ERROR type mismatch resolving
```
Comment (compiler-errors) [inline src/test/ui/impl-trait/issues/issue-78722.rs:11]: lol, why is this error showing up all of the sudden?

Comment (oli-obk) [reply src/test/ui/impl-trait/issues/issue-78722.rs:11]: Maybe it's the option being overwritten issue? Const qualifs are done before borrowck, so if borrowck succeeds but qualiffails, we never know

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/test/ui/impl-trait/issues/issue-78722.rs (hunk 5 above)
Comment (compiler-errors) [reply src/test/ui/impl-trait/issues/issue-78722.rs:11]: > Maybe it's the option being overwritten issue?

Like overwriting `Some(_)` with `None` by accident? I don't think so.

I actually think it might be due to the fact that, since we put the tainted field into the MIR body itself, we have to load the MIR body before checking its tainted field. This might cause us to do some more computations on error paths than we did before. If so, then it's for that same reason I had to [add this logic](https://github.com/rust-lang/rust/pull/93691/files#diff-ec8e8ba5fe5c35f6cf8a40478ee1e03c46188228b9fb7b1d927c186d6f97021bR248-R253) to prevent a regression in the test `src/test/ui/issues/issue-77919.rs`.

I don't think this test emitting this extra error is necessarily a bad thing, though?

Comment (oli-obk) [reply src/test/ui/impl-trait/issues/issue-78722.rs:11]: no it's not a bad thing indeed.

**Comment by compiler-errors** [comment]
@oli-obk can you retry the perf run? think it got canceled because I had to push a re-blessed test

**Comment by compiler-errors** [comment]
hmm, interesting that this has mixed results. looks like these regressions are mostly due to more typeck computations? maybe I do just need to lift this into a query like `query mir_is_tainted_by_errors(did) -> Result<(), ErrorReported>` that we do before we actually try loading the MIR itself.

**Comment by oli-obk** [comment]
Yea it's weird. Borrowck wasn't invoked at all in these incremental runs before this PR. Even if the `optimized_mir` query was invoked a lot. Maybe this is about the promoteds? We may have not invoked borrowck for promoted mir computations before, but are now. You can try this out by not invoking the borrowck query and (wrongly) claiming the mIR of promoteds is always fine

**Comment by compiler-errors** [comment]
Don't we want to run borrowck on promoteds -- otherwise wouldn't this code (which currently ICEs) then continue to be incorrectly const-evaluated?

```
fn foo() {
    let x = [1usize; {
        let s: &'static str;
        s.len()
    }];
}
```

My theory is that this might be due to additional query calls (like normalize, solving predicates) that happen during MIR optimization passes. The previous logic of `load_mir` was essentially: check if there are tyck errors, then load MIR. now since we put the tainted field into the body, we unconditionally load MIR, then check if there are tyck and borrowck errors with the body's tainted field. Since we load the MIR unconditionally, we're running MIR optimization passes, right? Those might invoke additional query operations that we see in the regressions above.

**Comment by oli-obk** [comment]
Well, the perf tests are all successfully compiling, so we never get an error from typeck. This means we always load the MIR anyway and never short circuit due to typeck errors in the benchmarks

**Comment by compiler-errors** [comment]
That is a good point, totally forgot that these benchmarks are not failing test.

Then I have absolutely no idea where these additional queries are coming from then.

**Comment by oli-obk** [comment]
Maybe `ensure` is magic in some way beyond not making a copy of the result? Like maybe it doesn't require rebuilding borrowck when optimized mir is made, even though the optimized mir query is rerun

**Comment by compiler-errors** [comment]
I actually thought `ensure` did the opposite, forcing queries to be re-calculated even if they're cached.

**Comment by compiler-errors** [comment]
Well, regardless, I have a rewrite where I tried your first suggestion and just pulled the borrowck result into a query that returns `Option<ErrorReported>` and which is `cache_on_disk_if { true }`. It actually works really nicely, so I'm gonna force-push it to this branch so we can perf-test it.

**Comment by compiler-errors** [comment]
Literally don't know what to do to not make this a perf regression, but also, shouldn't we expect to be doing more work here? 

Like the whole point of fixing this ICE is that we weren't borrowcking constants and now we are.

In any case, happy to revert back to the previous approach with the tainted field on the mir::Body or to keep this one. I am just kinda stumped in general about this PR perf-wise.

Thoughts, @oli-obk?

**Comment by jackh726** [comment]
Imo it's honestly not that bad of a perf regression. And as you said, we are doing more work. I think the perf regression here is justified.

**Comment by oli-obk** [comment]
Thanks for trying out the various ways to go about this. From the perf results the one with the field on the body looks like the best one to me. I agree we should take the hit, as it is the only way to prevent the ICEs reliably.

**Comment by compiler-errors** [comment]
Alright, reset back to the commit for that approach then :)

**Comment by compiler-errors** [comment]
rebased, an error message went away but it was definitely due to opaque types PR being reverted.

@rustbot ready

**Comment by oli-obk** [comment]
@rustbot label: +perf-regression-triaged

This fixes a bunch of compiler crashes that occurred because we weren't checking for borrowck errors before doing const eval. The additional work is entirely expected and other avenues did not pan out.
